bin/eslint.js

```diff
@@ -45,33 +45,72 @@ function buildIssueJson(message, path) {
   return JSON.stringify(issue);
 }
 
-// Uses glob to traverse code directory and find files to analyze,
-// excluding files passed in with by CLI config, and including only
-// files in the list of desired extensions
-function fileWalk(excludePaths, extensions){
-  var analysisFiles = [];
-  var allFiles = glob.sync("/code/**/**", {});
-
-  allFiles.forEach(function(file, i, a){
-    if(excludePaths.indexOf(file.split("/code/")[1]) < 0) {
-      if(!fs.lstatSync(file).isDirectory()) {
-        var extension = "." + file.split(".").pop();
-
-        if(extensions.indexOf(extension) >= 0) {
-          analysisFiles.push(file);
+function exclusionBasedFileListBuilder(excludePaths) {
+  // Uses glob to traverse code directory and find files to analyze,
+  // excluding files passed in with by CLI config, and including only
+  // files in the list of desired extensions.
+  //
+  // Deprecated style of file expansion, supported for users of the old CLI.
+  return function(extensions) {
+    var analysisFiles = [];
+    var allFiles = glob.sync("/code/**/**", {});
+
+    allFiles.forEach(function(file, i, a){
+      if(excludePaths.indexOf(file.split("/code/")[1]) < 0) {
+        if(!fs.lstatSync(file).isDirectory()) {
+          var extension = "." + file.split(".").pop();
+
+          if(extensions.indexOf(extension) >= 0) {
+            analysisFiles.push(file);
+          }
         }
       }
-    }
-  });
+    });
+
+    return analysisFiles;
+  };
+}
+
+function isFileWithMatchingExtension(file, extensions) {
+  var extension = "." + file.split(".").pop();
+  return (extensions.indexOf(extension) >= 0);
+}
 
-  return analysisFiles;
+function inclusionBasedFileListBuilder(includePaths) {
+  // Uses glob to expand the files and directories in includePaths, filtering
+  // down to match the list of desired extensions.
+  return function(extensions) {
+    var analysisFiles = [];
+
+    includePaths.forEach(function(fileOrDirectory, i) {
+      if ((/\/$/).test(fileOrDirectory)) {
+        // if it ends in a slash, expand and push
+        var filesInThisDirectory = glob.sync(
+          "/code/" + fileOrDirectory + "/**/**"
+        );
+        filesInThisDirectory.forEach(function(file, j){
+          if(!fs.lstatSync(file).isDirectory()) {
```

Comment:
I've been bitten by this before. This will return true for symlinks, which will then probably cause an error. If there's an `isFile` check, we should use that.